api/v1alpha1/customnbimage_types.go

```diff
@@ -24,69 +24,102 @@ import (
-	CNBiPhaseResolving          = "Resolving"
-	CNBiPhaseRunning            = "Running"
-	CNBiPhaseBuilding           = "Building"
-	CNBiPhaseImporting          = "Importing"
-	CNBiPhaseOk                 = "Ready"
+	CNBiPhasePending            CNBiPhase = "Pending"
+	CNBiPhaseFailed             CNBiPhase = "Failed"
+	CNBiPhasePreparing          CNBiPhase = "Preparing"
+	CNBiPhaseCreatingRepository CNBiPhase = "CreatingRepository"
+	CNBiPhaseResolving          CNBiPhase = "Resolving"
+	CNBiPhaseRunning            CNBiPhase = "Running"
+	CNBiPhaseBuilding           CNBiPhase = "Building"
+	CNBiPhaseImporting          CNBiPhase = "Importing"
+	CNBiPhaseValidating         CNBiPhase = "Validating"
+	CNBiPhaseOk                 CNBiPhase = "Ready"
 )
 
-// Strategies we support for a CNBi
+// BuildType describes how to build a custom notebook image.
+// Only one of the following build types may be specified.
+// +kubebuilder:validation:MinLength:1
+// +kubebuilder:validation:Enum=ImageImport;PackageList;GitRepository
+type BuildType string
+
 const (
-	CNBiStrategyImageImport         = "import"
-	CNBiStrategyBuildUsingPython    = "build"
-	CNBiStrategyBuildUsingBaseImage = "baseImage"
+	// ImportImage will simply import the image from the given URL
+	ImportImage BuildType = "ImageImport"
+
+	// PackageList will build a custom image using a specific List of Python Packages
+	// if no RuntimeEnvironment is specified, a baseImage must be specified for the build
+	// TODO #70 this needs a validator to ensure that either a RuntimeEnvironment or a baseImage is specified
+	PackageList BuildType = "PackageList"
+
+	// BuildGitRepository will builds a custom image from a git repository
+	GitRepository BuildType = "GitRepository"
 )
 
-// CustomNBAnnotations is a list of annotations that are added to the custom notebook image
+// CNBi Annotations is a list of annotations that are added to the custom notebook image
 const (
-	CNBiAnnotationName        = "opendatahub.io/notebook-image-name"
-	CNBiAnnotationDescription = "opendatahub.io/notebook-image-desc"
-	CNBiAnnotationCreator     = "opendatahub.io/notebook-image-creator"
+	CNBiNameAnnotationKey        = "opendatahub.io/notebook-image-name"
+	CNBiDescriptionAnnotationKey = "opendatahub.io/notebook-image-desc"
+	CNBiCreatorAnnotationKey     = "opendatahub.io/notebook-image-creator"
 )
 
-// CustomNBImageStrategy is the strategy super-set of configurations for all strategies.
-type CustomNBImageStrategy struct {
-	// Type is the strategy
-	Type string `json:"type,omitempty"`
-	// From is the reference to the source image, used for import strategy
-	From string `json:"from,omitempty"`
+// BuildTypeSpec is the strategy super-set of configurations for all strategies.
+type BuildTypeSpec struct {
+	// BuildType is the strategy
+	// +required
+	// +kubebuilder:Required
+	BuildType BuildType `json:"buildType"`
+	// FromUrl is the reference to the source image, used for import strategy
```

Comment:
leftover "FromUrl":

```suggestion
	// FromImage is the reference to the source image, used for import strategy
```

But beyond that, a question/proposal:  what about having a single field for image references (say, `baseImage`? or `imageRef`?) used by import and build?

Reply:
I like `imageRef` but I am unsure if an imageRef to a builder image, like and s2i, is the same semantics/mechanics as an ubi8-py38? hmm, I think it is, right?

Reply:
Actually, `from` (not `fromImage`) would be more consistent with `BuildConfig`s and s2i:

https://docs.openshift.com/container-platform/4.11/cicd/builds/creating-build-inputs.html

Note though that in `BuildConfigs`, *from* is an object reference (can be an imagestream, imagestreamtag, dockerimage...) here we can just stick to a string for simplicity.

So, let's keep `fromImage`